File: 剑指offer/52. 两个链表的第一个公共节点.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        node1, node2 = headA, headB
        while node1!=node2:
            if node1:
                node1 = node1.next
            else:
                node1 = headB
            if node2:
                node2 = node2.next
            else:
                node2 = headA
        return node1
#时间复杂度：O(M+N)
#空间复杂度：O(1)
        
############或者简写如下
 class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        node1, node2 = headA, headB
        
        while node1 != node2:
            node1 = node1.next if node1 else headB
            node2 = node2.next if node2 else headA

        return node1

https://leetcode-cn.com/problems/liang-ge-lian-biao-de-di-yi-ge-gong-gong-jie-dian-lcof/solution/shuang-zhi-zhen-fa-lang-man-xiang-yu-by-ml-zimingm/

File: 剑指offer/C++_Version/52. 两个链表的第一个公共节点.cpp
//方法1
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) 
    {
        ListNode* node1 = headA;
        ListNode* node2 = headB;
        while(node1!=node2)
        {
            if(node1!=nullptr)
                node1 = node1->next;
            else
                node1 = headB;
            if(node2!=nullptr)
                node2 = node2->next;
            else
                node2 = headA;
        }    
        return node1;
    }
};
